src/core/components/pin-set.js

```diff
@@ -0,0 +1,275 @@
+'use strict'
+
+const multihashes = require('multihashes')
+const CID = require('cids')
+const protobuf = require('protocol-buffers')
+const crypto = require('crypto')
+const fnv1a = require('fnv1a')
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const varint = require('varint')
+
+const emptyKeyHash = 'QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n'
+const emptyKey = multihashes.fromB58String(emptyKeyHash)
+const defaultFanout = 256
+const maxItems = 8192
+
+// Protobuf interface
+const pbSchema = (
+  // from go-ipfs/pin/internal/pb/header.proto
+  'message Set { ' +
+    // 1 for now
+    'optional uint32 version = 1; ' +
+    // how many of the links are subtrees
+    'optional uint32 fanout = 2; ' +
+    // hash seed for subtree selection, a random number
+    'optional fixed32 seed = 3; ' +
+  '}'
+)
+const pb = protobuf(pbSchema)
+function readHeader (rootNode) {
+  // rootNode.data should be a buffer of the format:
+  // < varint(headerLength) | header | itemData... >
+  const rootData = rootNode.data
+  const hdrLength = varint.decode(rootData)
+  const vBytes = varint.decode.bytes
+  if (vBytes <= 0) {
+    return { err: 'Invalid Set header length' }
+  }
+  if (vBytes + hdrLength > rootData.length) {
+    return { err: 'Impossibly large set header length' }
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    return { err: 'Unsupported Set version: ' + header.version }
+  }
+  if (header.fanout > rootNode.links.length) {
+    return { err: 'Impossibly large fanout' }
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
+    hasChild: (root, childhash, callback, _links, _checked, _seen) => {
+      // callback (err, has)
+      if (callback.fired) { return }
+      if (typeof childhash === 'object') {
+        childhash = multihashes.toB58String(childhash)
+      }
+      _links = _links || root.links.length
+      _checked = _checked || 0
+      _seen = _seen || {}
+
+      if (!root.links.length && _links === _checked) {
+        // all nodes have been checked
+        return callback(null, false)
+      }
+      root.links.forEach((link) => {
+        const bs58link = multihashes.toB58String(link.multihash)
+        if (bs58link === childhash) {
+          callback.fired = true
+          return callback(null, true)
+        }
+        dag.get(new CID(link.multihash), (err, res) => {
+          if (err) {
+            callback.fired = true
```

Comment:
Where is this callback.fired used? You can always use https://www.npmjs.com/package/once to ensure it only gets called once.